e2e/shared/e2e.go

```diff
@@ -0,0 +1,377 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package shared
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"k8s.io/apimachinery/pkg/types"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/clientutil"
+	"github.com/cloudoperators/greenhouse/pkg/lifecycle"
+)
+
+const (
+	AdminKubeConfigPathEnv     = "GREENHOUSE_ADMIN_KUBECONFIG"
+	RemoteKubeConfigPathEnv    = "GREENHOUSE_REMOTE_KUBECONFIG"
+	remoteIntKubeConfigPathEnv = "GREENHOUSE_REMOTE_INT_KUBECONFIG"
+	ControllerLogsPathEnv      = "CONTROLLER_LOGS_PATH"
+	managerDeploymentName      = "greenhouse-controller-manager"
+	managerDeploymentNamespace = "greenhouse"
+	remoteExecutionEnv         = "EXECUTION_ENV"
+	realCluster                = "GARDENER"
+)
+
+type EClient string
+
+const (
+	AdminClient      EClient = "AdminClient"
+	RemoteClient     EClient = "RemoteClient"
+	AdminRESTClient  EClient = "AdminRESTClient"
+	RemoteRESTClient EClient = "RemoteRESTClient"
+	AdminClientSet   EClient = "AdminClientSet"
+	RemoteClientSet  EClient = "RemoteClientSet"
+)
+
+var defaultElapsedTime = 180 * time.Second
+
+type WaitApplyFunc func(resource lifecycle.RuntimeObject) error
+
+type clientBox struct {
+	client     client.Client
+	restClient *clientutil.RestClientGetter
+	clientSet  kubernetes.Interface
+}
+
+type TestEnv struct {
+	adminClusterClient    *clientBox
+	remoteClusterClient   *clientBox
+	TestNamespace         string
+	IsRealCluster         bool
+	RemoteKubeConfigBytes []byte
+}
+
+func NewExecutionEnv(userScheme ...func(s *runtime.Scheme) error) *TestEnv {
+	adminClusterClient, err := prepareClients(AdminKubeConfigPathEnv, userScheme...)
+	Expect(err).NotTo(HaveOccurred(), "error preparing admin cluster client")
+	remoteClusterClient, err := prepareClients(RemoteKubeConfigPathEnv)
+	Expect(err).NotTo(HaveOccurred(), "error preparing remote cluster client")
+
+	isReal := isRealCluster()
+	var remoteKubeCfgBytes []byte
+	var remoteKubeCfgPath string
+	if isReal {
```

Comment:
Just below in return struct IsRealCluster: isReal is added so that it can be used in tests if at all there is a logic needed to do something when it is a local cluster, hence variable usage.